cpp/src/arrow/util/iterator_test.cc

```diff
@@ -726,6 +730,142 @@ TEST(TestAsyncUtil, CompleteBackgroundStressTest) {
+  };
+  struct Callback {
+    void operator()(const Result<T>& result) { state_->in.store(false); }
+    std::shared_ptr<State> state_;
+  };
+
+  std::shared_ptr<State> state_;
+};
+
+TEST(TestAsyncUtil, SerialReadaheadSlowProducer) {
+  AsyncGenerator<TestInt> it = BackgroundAsyncVectorIt({1, 2, 3, 4, 5});
+  SlowSourcePreventingReentrant<TestInt> tracker(std::move(it));
+  SerialReadaheadGenerator<TestInt> serial_readahead(tracker, 2);
+  AssertAsyncGeneratorMatch({1, 2, 3, 4, 5},
+                            static_cast<AsyncGenerator<TestInt>>(serial_readahead));
+  tracker.AssertValid();
+}
+
+TEST(TestAsyncUtil, SerialReadaheadSlowConsumer) {
+  int num_delivered = 0;
+  auto source = [&num_delivered]() {
+    if (num_delivered < 5) {
+      return Future<TestInt>::MakeFinished(num_delivered++);
+    } else {
+      return Future<TestInt>::MakeFinished(IterationTraits<TestInt>::End());
+    }
+  };
+  SerialReadaheadGenerator<TestInt> serial_readahead(std::move(source), 3);
+  SleepABit();
+  ASSERT_EQ(0, num_delivered);
+  ASSERT_FINISHES_OK_AND_ASSIGN(auto next, serial_readahead());
+  ASSERT_EQ(0, next.value);
+  ASSERT_EQ(3, num_delivered);
+  AssertAsyncGeneratorMatch({1, 2, 3, 4},
+                            static_cast<AsyncGenerator<TestInt>>(serial_readahead));
+}
+
+TEST(TestAsyncUtil, SerialReadaheadStress) {
+  constexpr int NTASKS = 20;
+  constexpr int NITEMS = 50;
+  constexpr int EXPECTED_SUM = (NITEMS * (NITEMS - 1)) / 2;
+  for (int i = 0; i < NTASKS; i++) {
+    AsyncGenerator<TestInt> it = BackgroundAsyncVectorIt(RangeVector(NITEMS));
+    SerialReadaheadGenerator<TestInt> serial_readahead(it, 2);
+    unsigned int sum = 0;
+    auto visit_fut = VisitAsyncGenerator<TestInt>(
+        serial_readahead, [&sum](TestInt test_int) -> Status {
+          sum += test_int.value;
+          // Normally sleeping in a visit function would be a faux-pas but we want to slow
+          // the reader down to match the producer to maximize the stress
+          std::this_thread::sleep_for(kYieldDuration);
+          return Status::OK();
+        });
+    ASSERT_FINISHES_OK(visit_fut);
+    ASSERT_EQ(EXPECTED_SUM, sum);
+  }
+}
+
+TEST(TestAsyncUtil, SerialReadaheadStressFast) {
+  constexpr int NTASKS = 20;
+  constexpr int NITEMS = 50;
+  constexpr int EXPECTED_SUM = (NITEMS * (NITEMS - 1)) / 2;
+  for (int i = 0; i < NTASKS; i++) {
+    AsyncGenerator<TestInt> it = BackgroundAsyncVectorIt(RangeVector(NITEMS), false);
+    SerialReadaheadGenerator<TestInt> serial_readahead(it, 2);
+    unsigned int sum = 0;
+    auto visit_fut = VisitAsyncGenerator<TestInt>(serial_readahead,
+                                                  [&sum](TestInt test_int) -> Status {
+                                                    sum += test_int.value;
+                                                    return Status::OK();
+                                                  });
+    ASSERT_FINISHES_OK(visit_fut);
+    ASSERT_EQ(EXPECTED_SUM, sum);
```

Comment:
IIUC, the intent of this test is to assert that concurrent writes haven't clobbered each other in this sum, thus demonstrating that the visitor was never called re-entrantly. If that's the case, please instead assert that a mutex isn't locked or similar

Reply:
It's more to check for race conditions.  The background thread filling callbacks and the visitor thread pulling the serial readahead both share a pointer to State.  Both of these tests were pretty useful in exposing issues with the old AddCallback version.  However, I added a reentrancy check in there for good measure and got rid of the sum stuff.